spark/dl/src/main/scala/com/intel/analytics/bigdl/optim/ParallelOptimizer.scala

```diff
@@ -0,0 +1,1053 @@
+/*
+ * Copyright 2016 The BigDL Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.intel.analytics.bigdl.optim
+
+import com.intel.analytics.bigdl.{Module, _}
+import com.intel.analytics.bigdl.dataset._
+import com.intel.analytics.bigdl.nn.{Container, Graph, Module, Sequential, Utils}
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.utils._
+import java.io.{File, FilenameFilter}
+import java.text.SimpleDateFormat
+import java.util.Calendar
+
+import com.intel.analytics.bigdl.models.utils.ModelBroadcast
+import com.intel.analytics.bigdl.nn.mkldnn.MklDnnContainer
+import com.intel.analytics.bigdl.nn.mkldnn.Phase.TrainingPhase
+import com.intel.analytics.bigdl.optim.DistriOptimizer._
+import org.apache.commons.lang.exception.ExceptionUtils
+import com.intel.analytics.bigdl.visualization.{TrainSummary, ValidationSummary}
+import org.apache.log4j.Logger
+import org.apache.spark.TaskContext
+import org.apache.spark.rdd.{RDD, ZippedPartitionsWithLocalityRDD}
+
+import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
+import scala.concurrent.Future
+import scala.reflect.{ClassTag, classTag}
+
+object ParallelOptimizer {
+  import Optimizer._
+
+  val logger: Logger = Logger.getLogger(getClass)
+
+  /**
+   * Train the model.
+   *
+   * @param dataset train dataset
+   * @param coresPerNode cores per node
+   * @param state state table
+   * @param endWhen trigger to stop training
+   * @param metrics metrics
+   * @param models cached models
+   * @param optimMethods optimization methods
+   * @param validationTrigger validation trigger
+   * @param validationDataSet validation dataset
+   * @param validationMethods validation methods
+   * @param cacheTrigger cache trigger
+   * @param cachePath cache path
+   * @param trainSummary train summary
+   * @param validationSummary validation summary
+   * @param isOverWrite if overwrite the checkpoint
+   */
+  private[optim] def optimize[T: ClassTag](
```

Comment:
Should we add some unit tests for this important method, including local, spark local, spark distributed?

Reply:
Added, this is only for spark not for local mode